test/signalrclient-e2e-tests/hub_connection_tests.cpp

```diff
@@ -0,0 +1,184 @@
+{
+    auto hub_conn = std::make_shared<signalr::hub_connection>(url + U("SignalR"));
+    auto message = std::make_shared<utility::string_t>();
+    auto received_event = std::make_shared<pplx::event>();
+
+    auto hub_proxy = hub_conn->create_hub_proxy(U("hubConnection"));
+
+    hub_proxy.on(U("displayMessage"), [message, received_event](const web::json::value& arguments)
+    {
+        *message = arguments.serialize();
+        received_event->set();
+    });
+
+    hub_conn->start().then([&hub_proxy]()
+    {
+        web::json::value obj{};
+        obj[0] = web::json::value(U("test"));
+
+        return hub_proxy.invoke<web::json::value>(U("displayMessage"), obj);
+
+    }).get();
+
+    ASSERT_FALSE(received_event->wait(200));
+
+    ASSERT_EQ(*message, U("[\"Send: test\"]"));
+}
+
+TEST(hub_connection_tests, send_message_return)
+{
+    auto hub_conn = std::make_shared<signalr::hub_connection>(url + U("SignalR"));
+
+    auto hub_proxy = hub_conn->create_hub_proxy(U("hubConnection"));
+
+    auto test = hub_conn->start().then([&hub_proxy]()
+    {
+        web::json::value obj{};
+        obj[0] = web::json::value(U("test"));
+
+        return hub_proxy.invoke<web::json::value>(U("returnMessage"), obj);
+
+    }).get();
+
+    ASSERT_EQ(test.serialize(), U("\"test\""));
+}
+
+TEST(hub_connection_tests, send_message_after_connection_restart)
+{
+    auto hub_conn = std::make_shared<signalr::hub_connection>(url + U("SignalR"));
+    auto message = std::make_shared<utility::string_t>();
+    auto received_event = std::make_shared<pplx::event>();
+
+    auto hub_proxy = hub_conn->create_hub_proxy(U("hubConnection"));
+
+    hub_proxy.on(U("displayMessage"), [message, received_event](const web::json::value& arguments)
+    {
+        *message = arguments.serialize();
+        received_event->set();
+    });
+
+    hub_conn->start().get();
+
+    hub_conn->stop().get();
+
+    hub_conn->start().then([&hub_proxy]()
+    {
+        web::json::value obj{};
+        obj[0] = web::json::value(U("test"));
+
+        return hub_proxy.invoke<web::json::value>(U("displayMessage"), obj);
+
+    }).get();
+
+    ASSERT_FALSE(received_event->wait(200));
+
+    ASSERT_EQ(*message, U("[\"Send: test\"]"));
+}
+
+TEST(hub_connection_tests, send_message_after_reconnect)
+{
+    auto hub_conn = std::make_shared<signalr::hub_connection>(url + U("SignalR"));
+    auto message = std::make_shared<utility::string_t>();
+    auto reconnected_event = std::make_shared<pplx::event>();
+    auto received_event = std::make_shared<pplx::event>();
+
+    auto hub_proxy = hub_conn->create_hub_proxy(U("hubConnection"));
+
+    hub_conn->set_reconnected([reconnected_event]()
+    {
+        reconnected_event->set();
+    });
+
+    hub_proxy.on(U("displayMessage"), [message, received_event](const web::json::value& arguments)
+    {
+        *message = arguments.serialize();
+        received_event->set();
+    });
+
+    hub_conn->start().get();
+
+    try
```

Comment:
"forceReconnect" hub method throw to force reconnect, so here need to catch and test continue.